File: plugins/sd/hooks/assets/statusline.ts
#!/usr/bin/env bun
/**
 * Claude Code statusline: folder | model | context% | 5h%(time) | 7d%(time) | $extra
 *
 * $extra 는 ~/.claude/.credentials.json 의 OAuth accessToken 으로 Anthropic 공식 사용량
 * 엔드포인트(api.anthropic.com/api/oauth/usage)만 조회해 표시합니다. 본인 사용량 조회
 * 용도이며 토큰을 제3자에게 전송·저장하지 않습니다.
 */

import { spawn } from "node:child_process";
import { mkdir, open, readFile, rename, stat, unlink, writeFile } from "node:fs/promises";
import type { FileHandle } from "node:fs/promises";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

const homeDir = getHomeDir();
const cacheFilePath = `${homeDir}/.claude/statusline-cache.json`;
const cacheDirPath = dirname(cacheFilePath);
const lockFilePath = `${homeDir}/.claude/statusline-cache.lock`;
const credentialsFilePath = `${homeDir}/.claude/.credentials.json`;
const fetchIntervalSeconds = 180;
const staleLockMilliseconds = 60_000;
const ownedLockEnvName = "SD_STATUSLINE_LOCK_OWNER";

type JsonRecord = Record<string, unknown>;

interface LockHandle {
  close(): Promise<void>;
  release(): Promise<void>;
}

function formatModel(modelId: string): string {
  const match = /^claude-(\w+)-(\d+)-(\d+)/.exec(modelId);
  if (!match) return modelId;

  const modelFamily = match[1] ?? "";
  const displayFamily = `${modelFamily.charAt(0).toUpperCase()}${modelFamily.slice(1).toLowerCase()}`;
  return `${displayFamily} ${match[2]}.${match[3]}`;
}

function formatRemaining(resetEpochSeconds: number): string {
  const deltaSeconds = resetEpochSeconds - currentEpochSeconds();
  if (deltaSeconds <= 0) return "0m";

  const totalMinutes = Math.trunc(deltaSeconds / 60);
  const days = Math.trunc(totalMinutes / (24 * 60));
  const hours = Math.trunc((totalMinutes % (24 * 60)) / 60);
  const minutes = totalMinutes % 60;

  if (days > 0) return `${days}d${hours}h`;
  if (hours > 0) return `${hours}h${minutes}m`;
  return `${minutes}m`;
}

async function readCache(): Promise<JsonRecord | undefined> {
  try {
    const data = JSON.parse(await readFile(cacheFilePath, "utf8")) as unknown;
    return asRecord(data);
  } catch {
    return undefined;
  }
}

function shouldFetch(cache: JsonRecord | undefined): boolean {
  if (!cache) return true;

  const lastFetchSeconds = asNumber(cache["last_fetch_ts"]) ?? 0;
  return currentEpochSeconds() - lastFetchSeconds > fetchIntervalSeconds;
}

async function acquireLock(): Promise<LockHandle | undefined> {
  await mkdir(cacheDirPath, { recursive: true });

  const fileHandle = await openLockFile();
  if (!fileHandle) return undefined;

  await fileHandle.writeFile(`${process.pid}\n`, "utf8");

  const close = async (): Promise<void> => {
    await fileHandle.close().catch(() => undefined);
  };

  return {
    close,
    async release(): Promise<void> {
      await close();
      await unlink(lockFilePath).catch(() => undefined);
    },
  };
}

async function openLockFile(): Promise<FileHandle | undefined> {
  try {
    return await open(lockFilePath, "wx", 0o600);
  } catch (error) {
    if (getErrorCode(error) !== "EEXIST") return undefined;
  }

  if (!(await removeStaleLockIfNeeded())) return undefined;

  try {
    return await open(lockFilePath, "wx", 0o600);
  } catch {
    return undefined;
  }
}

async function removeStaleLockIfNeeded(): Promise<boolean> {
  try {
    const lockStat = await stat(lockFilePath);
    if (Date.now() - lockStat.mtimeMs <= staleLockMilliseconds) return false;

    await unlink(lockFilePath);
    return true;
  } catch {
    return false;
  }
}

async function withExclusiveLock(callback: () => Promise<void> | void): Promise<boolean> {
  const lockHandle = await acquireLock();
  if (!lockHandle) return false;

  try {
    await callback();
    return true;
  } finally {
    await lockHandle.release();
  }
}

async function trySpawnFetch(version: string): Promise<void> {
  const lockHandle = await acquireLock().catch(() => undefined);
  if (!lockHandle) return;

  let spawned = false;
  try {
    await lockHandle.close();
    const childProcess = spawn(process.execPath, [fileURLToPath(import.meta.url), "--fetch", version], {
      detached: true,
      env: { ...process.env, [ownedLockEnvName]: "1" },
      stdio: "ignore",
    });
    childProcess.unref();
    spawned = true;
  } catch {
    // statusline should never crash Claude Code
  } finally {
    if (!spawned) await lockHandle.release();
  }
}

async function doFetch(version: string): Promise<void> {
  const ownsSpawnLock = process.env[ownedLockEnvName] === "1";
  try {
    if (ownsSpawnLock) {
      await doFetchLocked(version);
      return;
    }

    await withExclusiveLock(async () => {
      await doFetchLocked(version);
    });
  } catch {
    // statusline should never crash Claude Code
  } finally {
    if (ownsSpawnLock) await unlink(lockFilePath).catch(() => undefined);
  }
}

async function doFetchLocked(version: string): Promise<void> {
  const cache = await readCache();
  if (cache && !shouldFetch(cache)) return;

  try {
    const credentials = await readJsonFile(credentialsFilePath);
    const oauth = asRecord(credentials["claudeAiOauth"]);
    if (!oauth) throw new Error("missing claudeAiOauth");

    const token = asString(oauth["accessToken"]);
    if (!token) throw new Error("missing accessToken");

    const expiresAtMilliseconds = asNumber(oauth["expiresAt"]) ?? 0;
    if (expiresAtMilliseconds < Date.now()) {
      await writeCache(cache, "token_expired");
      return;
    }

    const usageData = await fetchUsage(token, version);
    const extraUsage = asRecord(usageData["extra_usage"]) ?? {};
    const newCache: JsonRecord = {
      "last_fetch_ts": currentEpochSeconds(),
      "extra_usage": {
        "is_enabled": asBoolean(extraUsage["is_enabled"]) ?? false,
        "used_credits": extraUsage["used_credits"],
      },
      "error": null,
    };

    await writeCacheAtomic(newCache);
  } catch (error) {
    await writeCache(cache, formatErrorMessage(error));
  }
}

async function fetchUsage(token: string, version: string): Promise<JsonRecord> {
  const abortController = new AbortController();
  const timeout = setTimeout(() => abortController.abort(), 15_000);

  try {
    const response = await fetch("https://api.anthropic.com/api/oauth/usage", {
      headers: {
        "Authorization": `Bearer ${token}`,
        "anthropic-beta": "oauth-2025-04-20",
        "User-Agent": `claude-code/${version}`,
      },
      signal: abortController.signal,
    });

    if (!response.ok) throw new Error(`HTTP ${response.status}`);

    const data = (await response.json()) as unknown;
    const usageData = asRecord(data);
    if (!usageData) throw new Error("invalid usage response");
    return usageData;
  } finally {
    clearTimeout(timeout);
  }
}

async function writeCache(oldCache: JsonRecord | undefined, errorMessage: string | null): Promise<void> {
  const data: JsonRecord = oldCache ? { ...oldCache } : {};
  data["last_fetch_ts"] = currentEpochSeconds();
  data["error"] = errorMessage;
  await writeCacheAtomic(data);
}

async function writeCacheAtomic(data: JsonRecord): Promise<void> {
  await mkdir(cacheDirPath, { recursive: true });

  const tempFilePath = `${cacheFilePath}.${process.pid}.${Date.now()}.tmp`;
  try {
    await writeFile(tempFilePath, JSON.stringify(data), "utf8");
    await rename(tempFilePath, cacheFilePath);
  } catch (error) {
    await unlink(tempFilePath).catch(() => undefined);
    throw error;
  }
}

function formatRateLimit(rateLimit: JsonRecord | undefined): string {
  const usedPercentage = asNumber(rateLimit?.["used_percentage"]);
  const resetsAt = asNumber(rateLimit?.["resets_at"]);

  if (usedPercentage !== undefined && resetsAt !== undefined) {
    return `${formatPercentage(usedPercentage)}%(${formatRemaining(resetsAt)})`;
  }
  if (usedPercentage !== undefined) return `${formatPercentage(usedPercentage)}%`;
  return "?";
}

async function main(): Promise<void> {
  const stdinData = await readStdinJson();

  const workspace = asRecord(stdinData["workspace"]);
  const currentDir = asString(workspace?.["current_dir"]) ?? asString(stdinData["cwd"]) ?? "";
  const folder = currentDir ? basename(currentDir) : "?";

  const modelData = asRecord(stdinData["model"]);
  const modelId = asString(modelData?.["id"]) ?? "";
  let model = modelId ? formatModel(modelId) : "?";

  const effortLevel = asString(asRecord(stdinData["effort"])?.["level"]);
  if (effortLevel) model = `${model} ${effortLevel}`;

  const contextWindow = asRecord(stdinData["context_window"]) ?? {};
  const usedPercentage = contextWindow["used_percentage"];
  const contextText = usedPercentage !== undefined && usedPercentage !== null ? `${usedPercentage}%` : "?";

  const rateLimits = asRecord(stdinData["rate_limits"]) ?? {};
  const fiveHourText = formatRateLimit(asRecord(rateLimits["five_hour"]));
  const sevenDayText = formatRateLimit(asRecord(rateLimits["seven_day"]));

  const cache = await readCache();
  const extraText = formatExtraUsage(cache);

  const version = asString(stdinData["version"]) ?? "2.1.86";
  if (shouldFetch(cache)) await trySpawnFetch(version);

  const parts = [folder, model, contextText, fiveHourText, sevenDayText];
  if (extraText) parts.push(extraText);
  console.log(parts.join(" | "));
}

function formatExtraUsage(cache: JsonRecord | undefined): string {
  if (!cache || cache["error"]) return "";

  const extraUsage = asRecord(cache["extra_usage"]);
  if (!extraUsage) return "";

  const isEnabled = extraUsage["is_enabled"] === true;
  const usedCredits = asNumber(extraUsage["used_credits"]);
  if (!isEnabled || usedCredits === undefined) return "";

  return `$${(usedCredits / 100).toFixed(2)}`;
}

async function readJsonFile(filePath: string): Promise<JsonRecord> {
  const data = JSON.parse(await readFile(filePath, "utf8")) as unknown;
  const record = asRecord(data);
  if (!record) throw new Error(`invalid JSON object: ${filePath}`);
  return record;
}

async function readStdinJson(): Promise<JsonRecord> {
  try {
    if (process.stdin.isTTY) return {};

    const chunks: Buffer[] = [];
    for await (const chunk of process.stdin) {
      chunks.push(typeof chunk === "string" ? Buffer.from(chunk) : chunk);
    }

    const text = Buffer.concat(chunks).toString("utf8").trim();
    if (!text) return {};

    const data = JSON.parse(text) as unknown;
    return asRecord(data) ?? {};
  } catch {
    return {};
  }
}

function basename(filePath: string): string {
  const normalizedPath = filePath.replace(/\\/g, "/");
  const parts = normalizedPath.split("/").filter(Boolean);
  return parts.at(-1) ?? "";
}

function formatPercentage(value: number): string {
  return Number.parseFloat(value.toFixed(2)).toString();
}

function currentEpochSeconds(): number {
  return Date.now() / 1000;
}

function getHomeDir(): string {
  const candidateHomeDir = process.env["HOME"] ?? process.env["USERPROFILE"];
  return candidateHomeDir || dirname(fileURLToPath(import.meta.url));
}

function asRecord(value: unknown): JsonRecord | undefined {
  return typeof value === "object" && value !== null && !Array.isArray(value) ? (value as JsonRecord) : undefined;
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" ? value : undefined;
}

function asNumber(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function asBoolean(value: unknown): boolean | undefined {
  return typeof value === "boolean" ? value : undefined;
}

function getErrorCode(error: unknown): string | undefined {
  const code = asRecord(error)?.["code"];
  return typeof code === "string" ? code : undefined;
}

function formatErrorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

try {
  const fetchArgIndex = process.argv.indexOf("--fetch");
  if (fetchArgIndex >= 0) {
    await doFetch(process.argv[fetchArgIndex + 1] ?? "2.1.86");
  } else {
    await main();
  }
} catch {
  // statusline should never crash Claude Code
}

File: plugins/sd/hooks/assets/statusline.py
#!/usr/bin/env python
"""Claude Code statusline: folder | model | context% | 5h%(time) | 7d%(time) | $extra

$extra 는 ~/.claude/.credentials.json 의 OAuth accessToken 으로 Anthropic 공식 사용량
엔드포인트(api.anthropic.com/api/oauth/usage)만 조회해 표시합니다. 본인 사용량 조회
용도이며 토큰을 제3자에게 전송, 저장하지 않습니다.

사용량 조회는 별도 프로세스로 떼어내 백그라운드에서 수행한다 — statusline 렌더가
네트워크 대기로 늦어지지 않도록. 동시 실행은 배타 락으로 하나만 조회하게 막는다.
"""

from __future__ import annotations

import contextlib
import json
import os
import re
import subprocess
import sys
import time
import urllib.request

FETCH_INTERVAL_SECONDS = 180
STALE_LOCK_SECONDS = 60
OWNED_LOCK_ENV_NAME = "SD_STATUSLINE_LOCK_OWNER"
USAGE_URL = "https://api.anthropic.com/api/oauth/usage"
USAGE_TIMEOUT_SECONDS = 15
DEFAULT_VERSION = "2.1.86"


def home_dir() -> str:
    return (
        os.environ.get("HOME")
        or os.environ.get("USERPROFILE")
        or os.path.dirname(os.path.abspath(__file__))
    )


def cache_file_path() -> str:
    return os.path.join(home_dir(), ".claude", "statusline-cache.json")


def cache_dir_path() -> str:
    return os.path.dirname(cache_file_path())


def lock_file_path() -> str:
    return os.path.join(home_dir(), ".claude", "statusline-cache.lock")


def credentials_file_path() -> str:
    return os.path.join(home_dir(), ".claude", ".credentials.json")


# --- 표시 포맷 ---------------------------------------------------------------


def format_model(model_id: str) -> str:
    match = re.match(r"^claude-(\w+)-(\d+)-(\d+)", model_id)
    if not match:
        return model_id

    family = match.group(1)
    return f"{family[:1].upper()}{family[1:].lower()} {match.group(2)}.{match.group(3)}"


def format_remaining(reset_epoch_seconds: float) -> str:
    delta_seconds = reset_epoch_seconds - time.time()
    if delta_seconds <= 0:
        return "0m"

    total_minutes = int(delta_seconds // 60)
    days = total_minutes // (24 * 60)
    hours = (total_minutes % (24 * 60)) // 60
    minutes = total_minutes % 60

    if days > 0:
        return f"{days}d{hours}h"
    if hours > 0:
        return f"{hours}h{minutes}m"
    return f"{minutes}m"


def format_percentage(value: float) -> str:
    return f"{round(value, 2):g}"


def format_rate_limit(rate_limit: dict | None) -> str:
    used_percentage = as_number((rate_limit or {}).get("used_percentage"))
    resets_at = as_number((rate_limit or {}).get("resets_at"))

    if used_percentage is not None and resets_at is not None:
        return f"{format_percentage(used_percentage)}%({format_remaining(resets_at)})"
    if used_percentage is not None:
        return f"{format_percentage(used_percentage)}%"
    return "?"


def format_extra_usage(cache: dict | None) -> str:
    if not cache or cache.get("error"):
        return ""

    extra_usage = as_record(cache.get("extra_usage"))
    if not extra_usage:
        return ""

    used_credits = as_number(extra_usage.get("used_credits"))
    if extra_usage.get("is_enabled") is not True or used_credits is None:
        return ""

    return f"${used_credits / 100:.2f}"


# --- 캐시 -------------------------------------------------------------------


def read_cache() -> dict | None:
    try:
        with open(cache_file_path(), encoding="utf-8") as handle:
            return as_record(json.load(handle))
    except OSError, ValueError:
        return None


def should_fetch(cache: dict | None) -> bool:
    if not cache:
        return True
    return (
        time.time() - (as_number(cache.get("last_fetch_ts")) or 0)
        > FETCH_INTERVAL_SECONDS
    )


def write_cache_atomic(data: dict) -> None:
    os.makedirs(cache_dir_path(), exist_ok=True)

    temp_path = f"{cache_file_path()}.{os.getpid()}.{int(time.time() * 1000)}.tmp"
    try:
        with open(temp_path, "w", encoding="utf-8") as handle:
            json.dump(data, handle)
        os.replace(temp_path, cache_file_path())
    except OSError:
        with contextlib.suppress(OSError):
            os.remove(temp_path)
        raise


def write_cache(old_cache: dict | None, error_message: str | None) -> None:
    data = dict(old_cache) if old_cache else {}
    data["last_fetch_ts"] = time.time()
    data["error"] = error_message
    write_cache_atomic(data)


# --- 락 ---------------------------------------------------------------------


def acquire_lock() -> int | None:
    """배타 락 획득. 이미 잡혀 있으면 None. 60초 넘게 방치된 락은 걷어낸다."""
    os.makedirs(cache_dir_path(), exist_ok=True)

    descriptor = open_lock_file()
    if descriptor is None:
        return None

    with contextlib.suppress(OSError):
        os.write(descriptor, f"{os.getpid()}\n".encode())
    return descriptor


def open_lock_file() -> int | None:
    flags = os.O_CREAT | os.O_EXCL | os.O_WRONLY
    try:
        return os.open(lock_file_path(), flags, 0o600)
    except FileExistsError:
        pass
    except OSError:
        return None

    if not remove_stale_lock_if_needed():
        return None

    try:
        return os.open(lock_file_path(), flags, 0o600)
    except OSError:
        return None


def remove_stale_lock_if_needed() -> bool:
    try:
        if time.time() - os.stat(lock_file_path()).st_mtime <= STALE_LOCK_SECONDS:
            return False
        os.remove(lock_file_path())
        return True
    except OSError:
        return False


def close_lock(descriptor: int) -> None:
    with contextlib.suppress(OSError):
        os.close(descriptor)


def release_lock(descriptor: int) -> None:
    close_lock(descriptor)
    with contextlib.suppress(OSError):
        os.remove(lock_file_path())


# --- 사용량 조회 -------------------------------------------------------------


def try_spawn_fetch(version: str) -> None:
    """조회를 백그라운드 프로세스로 떼어낸다. 락 소유권은 환경변수로 자식에 넘긴다."""
    descriptor = acquire_lock()
    if descriptor is None:
        return

    spawned = False
    try:
        close_lock(descriptor)

        env = dict(os.environ)
        env[OWNED_LOCK_ENV_NAME] = "1"

        kwargs: dict = {
            "stdin": subprocess.DEVNULL,
            "stdout": subprocess.DEVNULL,
            "stderr": subprocess.DEVNULL,
            "env": env,
        }
        if os.name == "nt":
            kwargs["creationflags"] = (
                subprocess.DETACHED_PROCESS | subprocess.CREATE_NEW_PROCESS_GROUP
            )
        else:
            kwargs["start_new_session"] = True

        subprocess.Popen(  # noqa: S603
            [sys.executable, os.path.abspath(__file__), "--fetch", version], **kwargs
        )
        spawned = True
    except Exception:
        # statusline should never crash Claude Code
        pass
    finally:
        if not spawned:
            release_lock(descriptor)


def do_fetch(version: str) -> None:
    owns_spawn_lock = os.environ.get(OWNED_LOCK_ENV_NAME) == "1"
    try:
        if owns_spawn_lock:
            do_fetch_locked(version)
            return

        descriptor = acquire_lock()
        if descriptor is None:
            return
        try:
            do_fetch_locked(version)
        finally:
            release_lock(descriptor)
    except Exception:
        # statusline should never crash Claude Code
        pass
    finally:
        if owns_spawn_lock:
            with contextlib.suppress(OSError):
                os.remove(lock_file_path())


def do_fetch_locked(version: str) -> None:
    cache = read_cache()
    if cache and not should_fetch(cache):
        return

    try:
        with open(credentials_file_path(), encoding="utf-8") as handle:
            credentials = as_record(json.load(handle)) or {}

        oauth = as_record(credentials.get("claudeAiOauth"))
        if not oauth:
            raise ValueError("missing claudeAiOauth")

        token = oauth.get("accessToken")
        if not isinstance(token, str) or not token:
            raise ValueError("missing accessToken")

        expires_at_milliseconds = as_number(oauth.get("expiresAt")) or 0
        if expires_at_milliseconds < time.time() * 1000:
            write_cache(cache, "token_expired")
            return

        usage_data = fetch_usage(token, version)
        extra_usage = as_record(usage_data.get("extra_usage")) or {}
        write_cache_atomic(
            {
                "last_fetch_ts": time.time(),
                "extra_usage": {
                    "is_enabled": extra_usage.get("is_enabled") is True,
                    "used_credits": extra_usage.get("used_credits"),
                },
                "error": None,
            }
        )
    except Exception as error:
        write_cache(cache, str(error) or type(error).__name__)


def fetch_usage(token: str, version: str) -> dict:
    request = urllib.request.Request(  # noqa: S310
        USAGE_URL,
        headers={
            "Authorization": f"Bearer {token}",
            "anthropic-beta": "oauth-2025-04-20",
            "User-Agent": f"claude-code/{version}",
        },
    )

    with urllib.request.urlopen(request, timeout=USAGE_TIMEOUT_SECONDS) as response:  # noqa: S310
        payload = json.loads(response.read().decode("utf-8"))

    usage_data = as_record(payload)
    if usage_data is None:
        raise ValueError("invalid usage response")
    return usage_data


# --- 렌더 -------------------------------------------------------------------


def main() -> None:
    stdin_data = read_stdin_json()

    workspace = as_record(stdin_data.get("workspace"))
    current_dir = as_string((workspace or {}).get("current_dir")) or as_string(
        stdin_data.get("cwd")
    )
    folder = basename(current_dir) if current_dir else "?"

    model_id = as_string((as_record(stdin_data.get("model")) or {}).get("id"))
    model = format_model(model_id) if model_id else "?"

    effort_level = as_string((as_record(stdin_data.get("effort")) or {}).get("level"))
    if effort_level:
        model = f"{model} {effort_level}"

    used_percentage = (as_record(stdin_data.get("context_window")) or {}).get(
        "used_percentage"
    )
    context_text = f"{used_percentage}%" if used_percentage is not None else "?"

    rate_limits = as_record(stdin_data.get("rate_limits")) or {}
    five_hour_text = format_rate_limit(as_record(rate_limits.get("five_hour")))
    seven_day_text = format_rate_limit(as_record(rate_limits.get("seven_day")))

    cache = read_cache()
    extra_text = format_extra_usage(cache)

    if should_fetch(cache):
        try_spawn_fetch(as_string(stdin_data.get("version")) or DEFAULT_VERSION)

    parts = [folder, model, context_text, five_hour_text, seven_day_text]
    if extra_text:
        parts.append(extra_text)
    print(" | ".join(parts))


def read_stdin_json() -> dict:
    try:
        if sys.stdin is None or sys.stdin.isatty():
            return {}
        text = sys.stdin.read().strip()
        return (as_record(json.loads(text)) or {}) if text else {}
    except Exception:
        return {}


def basename(file_path: str) -> str:
    parts = [part for part in file_path.replace("\\", "/").split("/") if part]
    return parts[-1] if parts else ""


def as_record(value: object) -> dict | None:
    return value if isinstance(value, dict) else None


def as_string(value: object) -> str | None:
    return value if isinstance(value, str) else None


def as_number(value: object) -> float | None:
    return (
        float(value)
        if isinstance(value, (int, float)) and not isinstance(value, bool)
        else None
    )


if __name__ == "__main__":
    try:
        for stream in (sys.stdout, sys.stderr):
            if stream is not None:
                stream.reconfigure(encoding="utf-8")

        if "--fetch" in sys.argv:
            index = sys.argv.index("--fetch")
            do_fetch(
                sys.argv[index + 1] if index + 1 < len(sys.argv) else DEFAULT_VERSION
            )
        else:
            main()
    except Exception:
        # statusline should never crash Claude Code
        pass
